File: server/src/agent/base-agent.ts
import { Logger } from '@nestjs/common';
import { BaseMessage, HumanMessage } from '@langchain/core/messages';
import { AgentState } from './model/agent-state.enum';
import { ChatModel } from '../llm/chat-model';
import { Observable, Subject } from 'rxjs';
import { EventName } from './model/event-name.enum';

export abstract class BaseAgent {
  protected readonly logger = new Logger(this.constructor.name);

  name = this.constructor.name;
  systemPrompt = '';
  nextStepPrompt = '';
  state: AgentState = AgentState.IDLE;
  currentStep = 0;
  maxSteps = 10;
  chatModel: ChatModel | null = null;
  messages: BaseMessage[] = [];

  protected get isNotFinished() {
    return this.state !== AgentState.FINISHED;
  }

  async run(userPrompt: string): Promise<string> {
    if (this.state !== AgentState.IDLE && this.state !== AgentState.FINISHED) {
      throw new Error(`Cannot run agent from state: ${this.state}`);
    }
    if (!userPrompt.trim()) {
      throw new Error('Cannot run agent with empty user prompt');
    }
    this.state = AgentState.RUNNING;
    this.currentStep = 0;
    this.messages.push(new HumanMessage(userPrompt));
    const results: string[] = [];
    try {
      for (let i = 0; i < this.maxSteps && this.isNotFinished; i++) {
        const stepNumber = i + 1;
        this.currentStep = stepNumber;
        this.logger.log(`Executing step ${stepNumber}/${this.maxSteps}`);
        const stepResult = await this.step();
        const result = `Step ${stepNumber}: ${stepResult}`;
        results.push(result);
      }
      return results.join('\n');
    } catch (err) {
      this.state = AgentState.ERROR;
      this.logger.error('Executing agent error:', err);
      return 'Executing agent error';
    } finally {
      this.cleanup();
    }
  }

  runStream(userPrompt: string): Observable<string> {
    const subject = new Subject<string>();

    (async () => {
      try {
        if (
          this.state !== AgentState.IDLE &&
          this.state !== AgentState.FINISHED
        ) {
          subject.next(
            JSON.stringify({
              type: EventName.ERROR,
              content: `Cannot run agent from state: ${this.state}`,
            }),
          );
          subject.complete();
          return;
        }
        if (!userPrompt.trim()) {
          subject.next(
            JSON.stringify({
              type: EventName.ERROR,
              content: 'Cannot run agent with empty user prompt',
            }),
          );
          subject.complete();
          return;
        }
      } catch (err) {
        subject.error(err);
        return;
      }

      this.state = AgentState.RUNNING;
      this.currentStep = 0;
      this.messages.push(new HumanMessage(userPrompt));
      try {
        for (let i = 0; i < this.maxSteps && this.isNotFinished; i++) {
          const stepNumber = i + 1;
          this.currentStep = stepNumber;
          this.logger.log(`Executing step ${stepNumber}/${this.maxSteps}`);
          subject.next(
            JSON.stringify({
              type: EventName.STEP_START,
              step: stepNumber,
              max_steps: this.maxSteps,
            }),
          );
          const stepResult = await this.stepWithEmitter(subject);
          subject.next(
            JSON.stringify({
              type: EventName.STEP_RESULT,
              step: stepNumber,
              content: stepResult,
            }),
          );
        }
        if (this.currentStep >= this.maxSteps) {
          this.state = AgentState.FINISHED;
          subject.next(
            JSON.stringify({
              type: AgentState.FINISHED,
              content: `Terminated, reached max steps (${this.maxSteps})`,
            }),
          );
        }
        subject.complete();
      } catch (err) {
        this.state = AgentState.ERROR;
        this.logger.error('Executing agent error:', err);
        subject.next(
          JSON.stringify({
            type: EventName.ERROR,
            content: 'Executing agent error',
          }),
        );
        subject.complete();
      } finally {
        this.cleanup();
      }
    })().catch((err) => {
      this.logger.error('Stream error:', err);
    });

    return subject.asObservable();
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected async stepWithEmitter(emitter: Subject<string>): Promise<string> {
    return this.step();
  }

  abstract step(): Promise<string>;

  protected cleanup(): void {}
}

File: src/main/java/com/github/tianchenghang/agent/BaseAgent.java
package com.github.tianchenghang.agent;

import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONUtil;
import com.github.tianchenghang.agent.model.AgentState;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.chat.messages.Message;
import org.springframework.ai.chat.messages.UserMessage;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

@Data
@Slf4j
public abstract class BaseAgent {
  static class EventNames {
    static final String STEP_START = "step-start";
    static final String STEP_RESULT = "step-result";
    static final String FINISHED = "finished";
    static final String ERROR = "error";
    static final String THINKING = "thinking";
    static final String TOOL_CALL = "tool-call";
    static final String TOOL_RESULT = "tool-result";
  }

  private String name;
  private String systemPrompt;
  private String nextStepPrompt;
  private AgentState state = AgentState.IDLE;
  private int currentStep = 0;
  private int maxSteps = 10;
  private ChatClient chatClient;
  private List<Message> messageList = new ArrayList<>();
  protected SseEmitter sseEmitter;

  protected void sendSseMessage(String type, Map<String, Object> data) {
    if (sseEmitter != null) {
      try {
        var message = new HashMap<String, Object>();
        message.put("type", type);
        message.putAll(data);
        sseEmitter.send(JSONUtil.toJsonStr(message));
      } catch (IOException e) {
        log.error("Send SSE message error:", e);
      }
    }
  }

  protected void sendSseMessage(String type, String content) {
    var data = new HashMap<String, Object>();
    data.put("content", content);
    sendSseMessage(type, data);
  }

  public String run(String userPrompt) {
    if (this.state != AgentState.IDLE) {
      throw new RuntimeException("Cannot run agent from state: " + this.state);
    }
    if (StrUtil.isBlank(userPrompt)) {
      throw new RuntimeException("Cannot run agent with empty user prompt");
    }
    this.state = AgentState.RUNNING;
    messageList.add(new UserMessage(userPrompt));
    var results = new ArrayList<String>();
    try {
      for (var i = 0; i < maxSteps && state != AgentState.FINISHED; i++) {
        var stepNumber = i + 1;
        currentStep = stepNumber;
        log.info("Executing step {}/{}", stepNumber, maxSteps);
        var stepResult = step();
        var result = "Step " + stepNumber + ": " + stepResult;
        results.add(result);
      }
      if (currentStep >= maxSteps) {
        state = AgentState.FINISHED;
        results.add("Terminated, reached max steps (" + maxSteps + ")");
      }
      return String.join("\n", results);
    } catch (Exception e) {
      state = AgentState.ERROR;
      log.error("Executing agent error: ", e);
      return "Executing agent error: " + e.getMessage();
    } finally {
      this.cleanup();
    }
  }

  public abstract String step();

  protected void cleanup() {}

  public SseEmitter runStream(String userPrompt) {
    var emitter = new SseEmitter(300_000L);
    this.sseEmitter = emitter;

    CompletableFuture.runAsync(
        () -> {
          try {
            if (this.state != AgentState.IDLE) {
              sendSseMessage(EventNames.ERROR, "Cannot run agent from state: " + this.state);
              emitter.complete();
              return;
            }
            if (StrUtil.isBlank(userPrompt)) {
              sendSseMessage(EventNames.ERROR, "Cannot run agent with empty user prompt");
              emitter.complete();
              return;
            }
          } catch (Exception e) {
            emitter.completeWithError(e);
          }
          this.state = AgentState.RUNNING;
          messageList.add(new UserMessage(userPrompt));
          // var results = new ArrayList<String>();
          try {
            for (var i = 0; i < maxSteps && state != AgentState.FINISHED; i++) {
              var stepNumber = i + 1;
              currentStep = stepNumber;
              log.info("Executing step {}/{}", stepNumber, maxSteps);
              var stepStartData = new HashMap<String, Object>();
              stepStartData.put("step", stepNumber);
              stepStartData.put("max_steps", maxSteps);
              sendSseMessage(EventNames.STEP_START, stepStartData);
              var stepResult = step();
              // var result = "Step " + stepNumber + ": " + stepResult;
              // results.add(result);
              var stepResultData = new HashMap<String, Object>();
              stepResultData.put("step", stepNumber);
              stepResultData.put("content", stepResult);
              sendSseMessage(EventNames.STEP_RESULT, stepResultData);
            }
            if (currentStep >= maxSteps) {
              state = AgentState.FINISHED;
              // results.add("Terminated, reached max steps (" + maxSteps + ")");
              sendSseMessage(
                  EventNames.FINISHED, "Terminated, reached max steps (" + maxSteps + ")");
            }
            emitter.complete();
          } catch (Exception e) {
            state = AgentState.ERROR;
            log.error("Executing agent error: ", e);
            try {
              sendSseMessage(EventNames.ERROR, "Executing agent error: " + e.getMessage());
              emitter.complete();
            } catch (Exception e2) {
              emitter.completeWithError(e2);
            }
          } finally {
            this.sseEmitter = null;
            this.cleanup();
          }
        });

    emitter.onTimeout(
        () -> {
          this.state = AgentState.ERROR;
          this.sseEmitter = null;
          this.cleanup();
          log.warn("SSE connection timeout");
        });

    emitter.onCompletion(
        () -> {
          if (this.state == AgentState.RUNNING) {
            this.state = AgentState.FINISHED;
          }
          this.sseEmitter = null;
          this.cleanup();
          log.info("SSE connection completed");
        });
    return emitter;
  }
}
